File: EXAMPLES/rgpio/files.c
#include <stdio.h>
#include <stdlib.h>

#include <lgpio.h>
#include <rgpio.h>

/*
gcc -Wall -o files files.c -lrgpio

./files
*/

int sbc;
int h;
int s;
int i;
int mode;
int fails;
char data[2048];

void check(char *func, int exp, int got)
{
   if (exp != got)
      printf("FAIL: %s (expected %d, got %d)\n", func, exp, got);
   else
      printf("PASS: %s\n", func);
}

void check_no_error(char *func, int got)
{
   if (got < 0)
      printf("FAIL: %s (expected no error, got %d)\n", func, got);
   else
      printf("PASS: %s (%d)\n", func, got);
}

void check_error(char *func, int got)
{
   if (got < 0)
      printf("PASS: %s (%d)\n", func, got);
   else
      printf("FAIL: %s (expected error, got %d)\n", func, got);
}

void check_not_null(char *func, int got)
{
   if (got == 0)
      printf("FAIL: %s (expected name, got %d)\n", func, got);
   else
      printf("PASS: %s (%d)\n", func, got);
}

int main(int argc, char *argv[])
{
   sbc = rgpiod_start(NULL, NULL);

   if (sbc < 0)
   {
      printf("connection failed\n");
      exit(-1);
   }

   // open file for write, create if it doesn't exist

   h = file_open(sbc,
      "test.file", LG_FILE_WRITE+LG_FILE_CREATE+LG_FILE_TRUNC);
   check_no_error("file_open", h);

   // write some text

   s = file_write(sbc, h, "Now is the winter of our discontent\n", 36);
   check_no_error("file_write", s);

   s = file_write(sbc, h, "Made glorious summer by this son of York\n", 41);
   check_no_error("file_write", s);

   s = file_close(sbc, h);
   check_no_error("file_close", s);

   // open for read

   h = file_open(sbc, "test.file", LG_FILE_READ);
   check_no_error("file_open", h);

   s = file_read(sbc, h, data, 100); // read up to 100 characters from file
   check("file_read", 77, s);

   s = file_seek(sbc, h, 25, LG_FROM_START);
   check_no_error("file_seek", s);

   s = file_read(sbc, h, data, 100); // read up to 100 characters from file
   check("file_read", 52, s);

   s = file_seek(sbc, h, -25, LG_FROM_END);
   check_no_error("file_seek", s);

   s = file_read(sbc, h, data, 100); // read up to 100 charactrs from file
   check("file_read", 25, s);

   s = file_seek(sbc, h, -50, LG_FROM_END);
   check_no_error("file_seek", s);

   s = file_read(sbc, h, data, 5); // read up to 5 charactrs from file
   check("file_read", 5, s);

   file_seek(sbc, h, -20, LG_FROM_CURRENT);
   check_no_error("file_seek", s);

   s = file_read(sbc, h, data, 100); // read up to 100 charactrs from file
   check("file_read", 65, s);

   s = file_close(sbc, h);
   check_no_error("file_close", s);

   s = file_list(sbc, "test*", data, 1000);
   check_no_error("file_list", s);

   char *pat2[]={"/tmp/unlikely-file-name", "file34"};

   for (i=0; i<sizeof(pat2)/sizeof(char*); i++)
   {
      s = file_list(sbc, pat2[i], data, 1000);
      check("file_list", LG_NO_FILE_MATCH, s);
   }

   char *pat3[]={"/tmp/*", "test.*"};

   for (i=0; i<sizeof(pat3)/sizeof(char*); i++)
   {
      s = file_list(sbc, pat3[i], data, 1000);
      check_no_error("file_list", s);
   }

   // create a file

   h = file_open(sbc,
      "/tmp/unlikely-file", LG_FILE_WRITE+LG_FILE_CREATE+LG_FILE_TRUNC);
   check_no_error("file_open", h);

   s = file_close(sbc, h);
   check_no_error("file_close", s);

   rgpiod_stop(sbc);
}

File: EXAMPLES/py_rgpio/files.py
#!/usr/bin/env python
"""
files.py
2020-11-18
Public Domain

http://abyz.me.uk/lg/py_rgpio.html

./files.py
"""

import rgpio

def check(func, exp, got):
   if exp != got:
      print("FAIL: {} (expected {}, got {})".format(func, exp, got))
   else:
      print("PASS: {}".format(func))

def check_no_error(func, got):
   if got < 0:
      print("FAIL: {} (expected no error, got {})".format(func, got))
   else:
      print("PASS: {} ({})".format(func, got))

def check_error(func, got):
   if got < 0:
      print("PASS: {} ({})".format(func, got))
   else:
      print("FAIL: {} (expected error, got {})".format(func, got))

def check_not_null(func, got):
   if got == "":
      print("FAIL: {} (expected name, got {})".format(func, got))
   else:
      print("PASS: {} ({})".format(func, got))

sbc = rgpio.sbc()
if not sbc.connected:
   exit()

sbc.set_user("test1")

# open file for write, create if it doesn't exist

h = sbc.file_open(
   "test.file", rgpio.FILE_WRITE+rgpio.FILE_CREATE+rgpio.FILE_TRUNC)
check_no_error("file_open", h)

# write some text

s = sbc.file_write(h, "Now is the winter of our discontent\n")
check_no_error("file_write", s)

s = sbc.file_write(h, "Made glorious summer by this son of York\n")
check_no_error("file_write", s)

s = sbc.file_close(h)
check_no_error("file_close", s)

# open for read

h = sbc.file_open("test.file", rgpio.FILE_READ)
check_no_error("file_open", h)

(s, data) = sbc.file_read(h, 100) # read up to 100 characters from file
check("file_read", 77, s)

s = sbc.file_seek(h, 25, rgpio.FROM_START)
check_no_error("file_seek", s)

(s, data) = sbc.file_read(h, 100) # read up to 100 characters from file
check("file_read", 52, s)

s = sbc.file_seek(h, -25, rgpio.FROM_END)
check_no_error("file_seek", s)

(s, data) = sbc.file_read(h, 100) # read up to 100 charactrs from file
check("file_read", 25, s)

s = sbc.file_seek(h, -50, rgpio.FROM_END)
check_no_error("file_seek", s)

(s, data) = sbc.file_read(h, 5) # read up to 5 charactrs from file
check("file_read", 5, s)

sbc.file_seek(h, -20, rgpio.FROM_CURRENT)
check_no_error("file_seek", s)

(s, data) = sbc.file_read(h, 100) # read up to 100 charactrs from file
check("file_read", 65, s)

s = sbc.file_close(h)
check_no_error("file_close", s)

(s, data) = sbc.file_list("file*")
check("file_list", 32, s)

rgpio.exceptions = False

for pat in "/", "/tmp", "*", "TEST":
   (s, data) = sbc.file_list(pat)
   check("file_list (" + pat + ")", rgpio.NO_FILE_ACCESS, s)

for pat in "/tmp/unlikely_file_name", "file34":
   (s, data) = sbc.file_list(pat)
   check("file_list (" + pat + ")", rgpio.NO_FILE_MATCH, s)

for pat in "/tmp/*", "file.*":
   (s, data) = sbc.file_list(pat)
   check_no_error("file_list (" + pat + ")", s)

# create a file

h = sbc.file_open(
   "/tmp/unlikely_file", rgpio.FILE_WRITE+rgpio.FILE_CREATE+rgpio.FILE_TRUNC)
check_no_error("file_open", h)

s = sbc.file_close(h)
check_no_error("file_close", s)

expected = [
   rgpio.NO_FILE_ACCESS, 1,                 1, 1,
   rgpio.NO_FILE_ACCESS, rgpio.NO_FILE_ACCESS, 1, 1,
   rgpio.NO_FILE_ACCESS, rgpio.NO_FILE_ACCESS, 1, 1,
   rgpio.NO_FILE_ACCESS, rgpio.NO_FILE_ACCESS, 1, 1,
   rgpio.NO_FILE_ACCESS, rgpio.NO_FILE_ACCESS, 1, 1,
   rgpio.NO_FILE_ACCESS, rgpio.NO_FILE_ACCESS, 1, 1,
   rgpio.NO_FILE_ACCESS, rgpio.NO_FILE_ACCESS, 1, 1,
   rgpio.NO_FILE_ACCESS, rgpio.NO_FILE_ACCESS, 1, 1]


fails = 0

for mode in range(0, 32):
   h = sbc.file_open("/tmp/unlikely_file", mode)
   if h < 0:
      if h != expected[mode]:
         fails += 1
         print("A: for {} expected {}, got {}".format(mode, expected[mode], h))
   else:
      if expected[mode] < 0:
         fails += 1
         print("A: for {} expected ok, got {}".format(mode, h))
      sbc.file_close(h)
check("A: file_open", 0, fails)

expected = [
   rgpio.NO_FILE_ACCESS, rgpio.NO_FILE_ACCESS,
   rgpio.NO_FILE_ACCESS, rgpio.NO_FILE_ACCESS,
   rgpio.NO_FILE_ACCESS, rgpio.NO_FILE_ACCESS,
   rgpio.NO_FILE_ACCESS, rgpio.NO_FILE_ACCESS,
   rgpio.NO_FILE_ACCESS, rgpio.NO_FILE_ACCESS,
   rgpio.NO_FILE_ACCESS, rgpio.NO_FILE_ACCESS,
   rgpio.NO_FILE_ACCESS, rgpio.NO_FILE_ACCESS,
   rgpio.NO_FILE_ACCESS, rgpio.NO_FILE_ACCESS,
   rgpio.NO_FILE_ACCESS, rgpio.NO_FILE_ACCESS,
   rgpio.NO_FILE_ACCESS, rgpio.NO_FILE_ACCESS,
   rgpio.NO_FILE_ACCESS, rgpio.NO_FILE_ACCESS,
   rgpio.NO_FILE_ACCESS, rgpio.NO_FILE_ACCESS,
   rgpio.NO_FILE_ACCESS, rgpio.NO_FILE_ACCESS,
   rgpio.NO_FILE_ACCESS, rgpio.NO_FILE_ACCESS,
   rgpio.NO_FILE_ACCESS, rgpio.NO_FILE_ACCESS,
   rgpio.NO_FILE_ACCESS, rgpio.NO_FILE_ACCESS]

fails = 0

for mode in range(0, 32):
   h = sbc.file_open("unlikely_file", mode)
   if h < 0:
      if h != expected[mode]:
         fails += 1
         print("B: for {} expected {}, got {}".format(mode, expected[mode], h))
   else:
      if expected[mode] < 0:
         fails += 1
         print("B: for {} expected ok, got {}".format(mode, h))
      sbc.file_close(h)
check("B: file_open", 0, fails)

fails = 0

for pat in "./zzz", "././yyy", "\.\./xxx", "../yyy", "./yyy":
   h = sbc.file_open(pat, rgpio.FILE_WRITE+rgpio.FILE_CREATE+rgpio.FILE_TRUNC)
   if h != rgpio.NO_FILE_ACCESS:
      fails += 1
      print("FAIL: for {} expected no access".format(root+pat))
      sbc.file_close(h)
check("file_open", 0, fails)

rgpio.exceptions = True

sbc.stop()

